lib/wibox/widget/calendar.lua

```diff
@@ -0,0 +1,405 @@
+    -- Split a string by a given `separator`
+    local function split(str, sep)
+        local sep, fields = sep or ":", {}
+        local pattern = string.format("([^%s]+)", sep)
+        string.gsub(str, pattern, function(c) fields[#fields+1] = c end)
+        return fields
+    end
+
+    local dfield = split(date, " ")
+    if #dfield == 3 then
+        return tonumber(dfield[1]), tonumber(dfield[2]), tonumber(dfield[3])
+    elseif #dfield == 2 then
+        return nil, tonumber(dfield[1]), tonumber(dfield[2])
+    elseif #dfield == 1 then
+        return nil, nil, tonumber(dfield[1])
+    end
+    return nil, nil, nil
+end
+
+
+--- Make a textbox
+-- @tparam string text Text of the textbox
+-- @tparam string font Font of the text
+-- @tparam boolean center Center the text horizontally
+-- @treturn wibox.widget.textbox
+local function make_cell(text, font, center)
+    return base.make_widget_declarative {
+        markup = text,
+        align  = center and "center" or "right",
+        valign = 'center',
+        font   = font,
+        widget = textbox
+    }
+end
+
+--- Create a grid layout with the month calendar
+-- @tparam table props Table of calendar properties
+-- @tparam string date Date in the "%d %m %Y" format
+-- @treturn widget Grid layout
+local function create_month(props, date)
+    local day, month, year = parse_date(date)
+    local num_rows    = 8
+    local num_columns = props.week_numbers and 8 or 7
+
+    -- Create grid layout
+    local layout = base.make_widget_declarative{
+        layout      = grid,
+        expand      = true,
+        homogeneous = true,
+        spacing     = props.spacing,
+        forced_num_rows = num_rows,
+        forced_num_cols = num_columns,
+    }
+
+    local start_row    = 3
+    local start_column = num_columns - 6
+    local week_start   = props.start_sunday and 1 or 2
+    local d            = os.date("*t", os.time{year=year, month=month+1, day=0})
+    local month_days   = d.day
+    local column_fday  = (d.wday - d.day + 1 - week_start ) % 7
+
+    --local flags = {"header", "weekdays", "weeknumber", "normal", "focus"}
+    local t, i, j, w, flag, text, markup
+
+    -- Header
+    flag = "header"
+    t = os.time{year=year, month=month, day=1}
+    text = os.date(props.hide_year and "%B" or "%B %Y", t)
+    w = props.fn_embed(make_cell(text, props.font, flag=="yearheader" or flag=="header"), flag)
+    layout:add_widget_at(w, 1, 1, 1, num_columns)
+
+    -- Week days
+    i = 2
+    for j = start_column, num_columns do  -- columns
+        flag = "weekdays"
+        t = os.time{year=2006, month=1, day=j-start_column+week_start}
+        text = string.sub(os.date("%a", t), 1, 2)
+        w = props.fn_embed(make_cell(text, props.font, flag=="yearheader" or flag=="header"), flag)
+        layout:add_widget_at(w, i, j, 1, 1)
+    end
+
+    -- Days
+    i = start_row
+    j = column_fday + start_column
+    local current_week = nil
+    for d=1, month_days do
```

Comment:
luacheck complaint:

    lib/wibox/widget/calendar.lua:196:9: shadowing definition of variable d on line 168